### DISCUSSION TITLE
Optimize processing pipeline

### DISCUSSION THREAD

**Discussion by devorient** [ImageMagick/ImageMagick#5940] [Help] [answered]
### ImageMagick version

7.1.0-52

### Operating system, version and so on

alpine x86_64

### Description

Disclaimer: I am an image processing noob.

I am trying to convert an invoice PDF to text and extract data from the invoice using regex
I use ImageMagick to convert the PDF to an image and then run Tesseract to get the text. I had to add some processing stages to improve the image quality, and hence the text output.

Here's the command I have arrived to (adapted from an old thread which I can't find anymore):

```sh
convert \ 
  \( -units PixelsPerInch -density 600 file.pdf -alpha off -monochrome -quality 100 \) \
  \( -clone 0 -morphology close rectangle:1x100 -negate \) \
  \( -clone 0 -morphology close rectangle:100x1 -negate \) \
  \( -clone 1 -clone 2 -evaluate-sequence add \) \
  -delete 1,2 -compose plus -composite result.png
```

My issue now is that this turned out to be super slow (~30 seconds on my machine). I would really appreciate some help making this faster.

Here's a sample PDF [file.pdf](https://github.com/ImageMagick/ImageMagick/files/10348261/file.pdf)
and the tesseract command if you're curious (tesseract 5.2.0)

```sh
tesseract -l eng --oem 3 --psm 6 -c preserve_interword_spaces=1 -c textonly_pdf=1 result.png out
```

Thanks in advance

**Comment by fmw42** [comment, marked as answer]
When you specify a density of 600 that rasterizes the vector PDF to about 600/72=8.3x larger in each dimension. So about 70x the number of pixels per channel. So your PDF which may be large in dimensions in inches, might be huge when rasterized. You then keep it at that size for all the rest of your processing.  You likely can reduce the size after rasterizing and maintain quality.  I typically use 72*4=288 density and then resize by 1/4=25%. But you can rasterize at your 600 and resize to whatever pixel size you think you need for tesseract.  

Your file is 8.5x11 inches (and default 72 dpi).  At 600 dpi that would rasterize to 600*8.5=5100 by 600*11=6600. So 5100x6600 pixels.

**Reply by devorient** [reply]
This made it a lot faster, thanks :)
